### PR TITLE
NDEV-521 mempool replication

### PR DESCRIPTION
**Pull request by rozhkovdmitrii**

### CODE REVIEW

**Review by afalaleev — DISMISSED** [review]

File: proxy/mempool/mempool_service.py [hunk 1]
```diff
@@ -56,13 +66,20 @@ async def process_mp_request(self, mp_request: MPRequest) -> Any:
             return self._mempool.get_pending_tx_by_hash(pending_tx_by_hash_req.tx_hash)
         elif mp_request.type == MPRequestType.GetGasPrice:
             return self._mempool.get_gas_price()
-        self.error(f"Failed to process mp_request, unknown type: {mp_request.type}")
+        self.error(f"Failed to process mp_reqeust, unknown type: {mp_request.type}")
```
Comment (afalaleev) [inline proxy/mempool/mempool_service.py]: Typo: mp_reqeust -> mp_request

File: proxy/mempool/mempool_api.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 from dataclasses import dataclass
 from enum import IntEnum
 
-from typing import Any, Optional, List
+from typing import Any, Optional, List, Tuple
```
Comment (afalaleev) [inline proxy/mempool/mempool_api.py]: Tuple is not used

File: proxy/mempool/mempool_schedule.py [hunk 3]
```diff
@@ -28,13 +28,22 @@ def __len__(self) -> int:
...
         self._impl = self._SortedQueueImpl(lt_key_func)
         self._eq_key_func = eq_key_func
 
-    def __getitem__(self, index: int) -> SortedQueueItem:
-        return self._impl.queue[index]
+    def __getitem__(self, index: Union[int, slice]) -> SortedQueueItem:
+        if isinstance(index, int):
+            return self._impl.queue[index]
+        elif isinstance(index, slice):
+            sorted_queue = SortedQueue[SortedQueueItem, int, str](lt_key_func=lambda a: -a.nonce,eq_key_func=lambda a: a.signature)
```
Comment (afalaleev) [inline proxy/mempool/mempool_schedule.py]: Sorted queue is used not only for nonce-queue.

1. If you want to define type - `SortedQueue[SortedQueueItem, SortedQueueLtKey, SortedQueueEqKey](lt_key_func=self._impl._lt_key_func, eq_key_func=self._eq_key_func)`
2. Return type of function is wrong.

File: proxy/mempool/mempool_schedule.py [hunk 4]
```diff
@@ -134,6 +146,9 @@ def __init__(self, sender_address: Optional[str] = None) -> None:
             eq_key_func=lambda a: a.signature
         )
 
+    def __len__(self):
+        return len(self._tx_nonce_queue)
```
Comment (afalaleev) [inline proxy/mempool/mempool_schedule.py]: For what? You already have `get_queue_len()`

File: proxy/mempool/mempool_service.py [hunk 5]
```diff
@@ -38,11 +40,19 @@ def start(self):
...
+        try:
+            if issubclass(type(mp_request), (MPRequest,)):
+                return await self.process_mp_request(cast(MPRequest, mp_request))
+            elif issubclass(type(mp_request), (MaintenanceRequest,)):
+                return self.process_maintenance_request(cast(MaintenanceRequest, mp_request))
+            self.error(f"Failed to process mp_request, unknown type: {type(mp_request)}")
+        except Exception as err:
+            log_ctx = {"context": {"req_id": mp_request.req_id}}
+            err_tb = "".join(traceback.format_tb(err.__traceback__))
+            self.error(f"Failed to process maintenance request: {mp_request.command}, {err}, traceback: {err_tb}", extra=log_ctx)
```
Comment (afalaleev) [inline proxy/mempool/mempool_service.py]:
```
with logging_context(req_id=mp_request.req_id):
    self._on_exception(f"Failed to process maintenance request: {mp_request.command}", err)
```

File: proxy/mempool/mempool_service.py [hunk 6]
```diff
@@ -73,9 +90,13 @@ def run(self):
...
+            self._replicator = MemPoolReplicator(self._mempool)
             self.event_loop.run_forever()
         except Exception as err:
             self.error(f"Failed to run mempool_service: {err}")
 
     def on_resource_released(self, resource_id: int):
         self._mempool.on_resource_got_available(resource_id)
+
+    def on_operator_resource_released(self):
+        self._mempool.on_operator_resource_released()
```
Comment (afalaleev) [inline proxy/mempool/mempool_service.py]: This event is absent.

File: proxy/mempool/mempool_schedule.py [hunk 7]
```diff
@@ -441,3 +465,22 @@ def set_sender_state_tx_cnt_list(self, sender_tx_cnt_list: List[MPSenderTxCntDat
             #   and now should be included into the execution queue
             if not sender_pool.is_paused():
                 self._schedule_sender_pool(sender_pool)
+
+    def get_taking_out_txs_iterator(self) -> Iterator[Tuple[str, MPTxRequestList]]:
+        empty_pools = []
+        for tx_pool in self._sender_pool_queue:
```
Comment (afalaleev) [inline proxy/mempool/mempool_schedule.py]: `self._sender_pool_queue` has only senders that are ready to be executed.
The full list of senders is in the `self._sender_pool_dict`.

File: proxy/mempool/mempool_schedule.py [hunk 8]
```diff
@@ -210,6 +225,15 @@ def cancel_process_tx(self, tx: MPTxRequest) -> None:
         self._processing_tx.neon_tx_exec_cfg = tx.neon_tx_exec_cfg
         self._processing_tx = None
 
+    def take_out_txs(self) -> MPTxRequestList:
+        is_processing = self._processing_tx is not None
+        self.debug(f"Take out txs from sender pool: {self.sender_address}, count: {len(self._tx_nonce_queue)}, processing: {is_processing}")
+        _from = 1 if is_processing else 0
+        taken_out_txs = self._tx_nonce_queue[_from:]
```
Comment (afalaleev) [inline proxy/mempool/mempool_schedule.py:230]: You need `List[MPTxRequest]`, but you return the `Sorted Queue of the MPTxRequest`...

I understand that you have implemented the traditional python way for lists.

But, If you add the method `take_out_txs` that has a custom implementation for the custom case, it looks better to make it really custom 
- in the sorted queue implement logic that returns the List! (not sorted queue!) of items by dropping them from the internal list of the sorted queue. 
- the implementation will really simple - you make a copy of the `_impl.queue[from:]`, and `_impl.queue = _impl.queue[:from]`
- don't call pop method for txs inside the `take_out_txs`

Another way - is to keep the structure of MPSenderTxPool simple wo high level logic in it....

File: proxy/mempool/mempool_schedule.py [hunk 9]
```diff
@@ -441,3 +465,22 @@ def set_sender_state_tx_cnt_list(self, sender_tx_cnt_list: List[MPSenderTxCntDat
...
+            taken_out_txs = tx_pool.take_out_txs()
+            for tx in taken_out_txs:
+                self._tx_dict.pop(tx)
+            if tx_pool.is_empty():
+                empty_pools.append(tx_pool)
+            yield tx_pool.sender_address, taken_out_txs
+
+        for tx_pool in empty_pools:
+            self._remove_empty_sender_pool(tx_pool)
+            self._sender_pool_queue.pop(tx_pool)
```
Comment (afalaleev) [inline proxy/mempool/mempool_schedule.py]: This place has a lot of problems ....

Comment (afalaleev) [reply proxy/mempool/mempool_schedule.py]: If pool is not in the queue of the execution it raises an exception

**Review by afalaleev — CHANGES_REQUESTED** [review]
I'm sorry, I can return to the PR later to provide information how to resolve issues.
Currently it has a lot of problems.

File: proxy/mempool/mempool_schedule.py (hunk 3 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: OMG, of course you are right !

File: proxy/mempool/mempool_schedule.py (hunk 7 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: It's a crucial thing I've missed thank you

File: proxy/mempool/mempool_schedule.py (hunk 9 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: That is impossible case, Andrew, because I stuff the empty_pools four lines before

Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: But as you mentioned before not all the pools are in _sender_pool_queue

Comment (afalaleev) [reply proxy/mempool/mempool_schedule.py]: Sender pool has 3 states:
if pool.is_paused() -> paused_sender_set (+),  sender_pool_queue(-)
elif pool.is_processing() -> paused_sender_set(-), send_pool_queue(-)
else -> paused_sender_set(-), sender_pool_queue(+)

File: proxy/mempool/mempool_service.py (hunk 1 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_service.py]: solved

File: proxy/mempool/mempool_api.py (hunk 2 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_api.py]: solved

File: proxy/mempool/mempool_schedule.py (hunk 3 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: Reduced with generic parameters

File: proxy/mempool/mempool_schedule.py (hunk 4 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: Dropped away

File: proxy/mempool/mempool_service.py (hunk 5 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_service.py]: I'll see what I can do with it in the python-logged-groups in the future issues. Now it's self._on_exception

File: proxy/mempool/mempool_service.py (hunk 6 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_service.py]: Thank you, solved

File: proxy/mempool/mempool_schedule.py (hunk 7 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: solved, walk through the `_sender_pool_dict`

File: proxy/mempool/mempool_schedule.py (hunk 8 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py:230]: I prefer to not touch Generic Sorted Queue. I made a mess with the return  type and the actual value, thank you for remark

Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py:230]: solved

File: proxy/mempool/mempool_schedule.py (hunk 9 above)
Comment (rozhkovdmitrii) [reply proxy/mempool/mempool_schedule.py]: solved

**Comment by rozhkovdmitrii** [comment]
@afalaleev solved all remarks, do merging develop and tests

File: proxy/mempool/mempool_service.py (hunk 5 above)
Comment (afalaleev) [reply proxy/mempool/mempool_service.py]: Ok, lets do it later.

**Review by afalaleev — APPROVED** [review]

**Pull request merged**